plugins/search-backend-node/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -0,0 +1,296 @@
+            )
+          : Object.values(this.documentTypeFields).reduce(
+              (acc, it) => acc.concat(it),
+              [],
+            ),
+      ),
+    ];
+
+    if (searchableFields.length === 0) {
+      searchableFields = ['*'];
+    }
+
+    return {
+      elasticQueryBuilder: () => {
+        const filter = Object.entries(filters)
+          .filter(([_, value]) => Boolean(value))
+          .map(([key, value]: [key: string, value: any]) => {
+            if (['string', 'number', 'boolean'].includes(typeof value)) {
+              return esb.matchQuery(key, value.toString());
+            }
+            if (Array.isArray(value)) {
+              return esb
+                .boolQuery()
+                .should(value.map(it => esb.matchQuery(key, it.toString())));
+            }
+            this.logger.error(
+              'Failed to query, unrecognized filter type',
+              key,
+              value,
+            );
+            throw new Error(
+              'Failed to add filters to query. Unrecognized filter type',
+            );
+          });
+        const query = isBlank(term)
+          ? esb.matchAllQuery()
+          : esb
+              .multiMatchQuery(searchableFields, term)
+              .fuzziness('auto')
+              .minimumShouldMatch(1);
+        return (
+          esb
+            .requestBodySearch()
+            .query(esb.boolQuery().filter(filter).must([query]))
+            // TODO: Replace size limit with page cursor after pagination approach decided
+            // See: https://github.com/backstage/backstage/issues/6062
+            .size(100)
+        );
+      },
+      documentTypes: types,
+    };
+  }
+
+  setTranslator(translator: ElasticSearchQueryTranslator) {
+    this.translator = translator;
+  }
+
+  async index(type: string, documents: IndexableDocument[]): Promise<void> {
+    this.logger.info(
+      `Started indexing ${documents.length} documents for index ${type}`,
+    );
+    const startTimestamp = process.hrtime();
+    const alias = this.constructSearchAlias(type);
+    try {
+      const aliases = await this.elasticSearchClient.cat.aliases({
+        format: 'json',
+        name: alias,
+      });
+      const removableIndices = aliases.body.map(
+        (r: Record<string, any>) => r.index,
+      );
+      const index = this.constructIndexName(type, `${Date.now()}`);
+
+      await this.elasticSearchClient.indices.create({
+        index,
+      });
+      const result = await this.elasticSearchClient.helpers.bulk({
+        datasource: documents,
+        onDocument() {
+          return {
+            index: { _index: index },
+          };
+        },
+        refreshOnCompletion: index,
+      });
+
+      // Store the indexed fields for the automatically created index
+      // so they are available and can be looked up when querying
+      const fields = await this.elasticSearchClient.indices.getFieldMapping({
```

Comment:
The current interface of these engines doesn't support specific mappings for the document type so we are using ElasticSearch dynamic mapping creation based on the data we are indexing (in the bulk insertion call above). 

That means the only real truth of fields we can query against is within ElasticSearch and ideally we do want to have knowledge of those to be able to omit/add/filter etc. based on them. There are few alternative approaches we could take here:
1. Use the first document of the passed in array and its keys as the source of truth for this list.
  * This would make the state within this search engine to be populated a bit faster and usable near immediately after app startup. 
  * The (first) object might have missing fields that subsequent objects would have, thus making possible fields in subsequents objects unsearchable
2. Modify the interface of these classes to take in a "mapping" interface which would contain the fields and their types that need to be indexed by ElasticSearch
  * This would introduce strict typing and explicitly defined searchable fields. Making it certain that no extra fields are querying nor omitted from a query.
3. Keep as is, ElasticSearch will create dynamic mapping of fields based on the data it retrieves
  * This will delay the first bite to be able to search until the time that that specific dataset is indexed. 
  * The index and their fields will definitely match ElasticSearch index mappings

Reply:
The fourth option is of course to just use wildcard instead of fields in query in the default Backstage implementation of this search and leave stricter mapping implementation for integrators if they have a need to do that. We'll achieve the same results from users perspective without the unnecessary additional possibly complex logic to keep track and define fields/mappings without knowing specific implementations.

This is the approach I'll propose we'll take at least with this iteration.